nucypher/network/server.py

```diff
@@ -334,73 +336,92 @@ def reencrypt_via_rest(id_as_hex):
+        as a decentralized node. Otherwise, we use the map_id in
+        federated mode.
+        """
         if not this_node.federated_only:
             from nucypher.policy.collections import SignedTreasureMap as _MapClass
         else:
             from nucypher.policy.collections import TreasureMap as _MapClass
 
+        # Step 1: First, we verify the signature of the received treasure map.
+        # This step also deserializes the treasure map iff it's signed correctly.
         try:
-            treasure_map = _MapClass.from_bytes(bytes_representation=request.data, verify=True)
+            received_treasure_map = _MapClass.from_bytes(bytes_representation=request.data, verify=True)
         except _MapClass.InvalidSignature:
-            log.info("Bad TreasureMap HRAC Signature; not storing {}".format(treasure_map_id))
+            log.info(f"Bad TreasureMap HRAC Signature; not storing for HRAC {received_treasure_map._hrac.hex()}")
             return Response("This TreasureMap's HRAC is not properly signed.", status=401)
 
-        treasure_map_index = bytes.fromhex(treasure_map_id)
-
-        # First let's see if we already have this map.
-
-        try:
-            previously_saved_map = this_node.treasure_maps[treasure_map_index]
-        except KeyError:
-            pass # We don't have the map.  We'll validate and perhaps save it.
-        else:
-            if previously_saved_map == treasure_map:
-                return Response("Already have this map.", status=303)
-                # Otherwise, if it's a different map with the same ID, we move on to validation.
-
-        if treasure_map.public_id() == treasure_map_id:
-            do_store = True
+        # Additionally, we determine the map identifier from the type of node. 
+        # If the node is federated, we also set the expiration for a week.
+        if not this_node.federated_only:
+            map_identifier = received_treasure_map._hrac.hex()
         else:
-            return Response("Can't save a TreasureMap with this ID from you.", status=409)
+            map_identifier = received_treasure_map.public_id()
+            expiration_date = MayaDT.from_datetime(datetime.utcnow() + timedelta(days=7))
 
-        if do_store and not this_node.federated_only:
-            alice_checksum_address = this_node.policy_agent.contract.functions.getPolicyOwner(
-                treasure_map._hrac[:16]).call()
-            do_store = treasure_map.verify_blockchain_signature(checksum_address=alice_checksum_address)
+        # Step 2: Check if we already have the treasure map.
+        try:
+            with datastore.describe(TreasureMap, map_identifier) as stored_treasure_map:
+                if _MapClass.from_bytes(stored_treasure_map.treasure_map) == received_treasure_map:
+                    return Response("Already have this map.", status=303)
+        except RecordNotFound:
+            # This appears to be a new treasure map that we don't have!
+            pass
```

Comment:
Queries module will clean this up into *get_or_create*

Reply:
Not necessarily, there's checks that we need to perform before we create it. Additionally, I think it's good practice to read from on-chain only once there are no remaining checks to be done.